File: python/bond_verify/evidence.py
"""Evidence verification and construction (§3.3, §3.5). An evidence
artifact is a signed Assertion whose `source` ref pins the native source
bytes. Supported OSS source profiles: `fixture/1` (source bytes = J(facts))
and `bond.draft-store/1` (source bytes = J(Signed<DraftStoreRecord>)).
Native upstream profiles require an installed profile verifier; without
one the result is UNSUPPORTED_COMPOSITION."""

from __future__ import annotations

from typing import Any, Callable

from .canon import (
    BondError, H, J, PACKAGE_LIMITS, b64u_decode, b64u_encode, deep_equal,
    parse_json_bytes,
)
from .schema import v_assertion, v_blob, v_draft_store_record, v_signed
from .sign import Signer, sign_object, verify_signed


def blob_of(data: bytes, media_type: str) -> dict:
    return {
        "hash": H(data),
        "bytes": len(data),
        "media_type": media_type,
        "data": b64u_encode(data),
    }


def ref_of(b: dict) -> dict:
    return {"hash": b["hash"], "bytes": b["bytes"], "media_type": b["media_type"]}


def json_blob(x: Any) -> dict:
    return blob_of(J(x).encode("utf-8"), "application/json")


def parse_blob(b: dict) -> bytes:
    if b["data"] is None:
        raise BondError("INCOMPLETE", "blob bytes absent")
    raw = b64u_decode(b["data"])
    if len(raw) != b["bytes"] or H(raw) != b["hash"]:
        raise BondError("HASH_MISMATCH", "blob content")
    return raw


def verify_source(facts: dict, assertion: dict, source_bytes: bytes,
                  trust: dict, tenant_id: str,
                  source_verifiers: dict | None = None) -> None:
    """Verify native source bytes against asserted facts."""
    profile = assertion["source_profile"]
    if profile == "fixture/1":
        if J(facts).encode("utf-8") != source_bytes:
            raise BondError("PROJECTION_INVALID",
                            "fixture source does not match facts")
        return
    if profile == "bond.draft-store/1":
        parsed = v_signed(parse_json_bytes(source_bytes, PACKAGE_LIMITS),
                          "draft-store", v_draft_store_record)
        verify_signed("draft-store", parsed, trust, tenant_id)
        if (parsed["body"]["action_hash"] != assertion["action_hash"]
                or not deep_equal(parsed["body"]["facts"], facts)):
            raise BondError("PROJECTION_INVALID",
                            "draft-store record does not match assertion")
        return
    v = (source_verifiers or {}).get(profile)
    if v is None:
        raise BondError("UNSUPPORTED_COMPOSITION",
                        f"no verifier installed for source profile {profile}")
    v(facts, source_bytes, assertion, trust, tenant_id)


def verify_evidence(signed_assertion: dict, source_blob: dict, trust: dict,
                    tenant_id: str, expected_action_hash: str,
                    source_verifiers: dict | None = None) -> dict:
    """Full evidence verification against an action hash."""
    verify_signed("assertion", signed_assertion, trust, tenant_id)
    if signed_assertion["body"]["action_hash"] != expected_action_hash:
        raise BondError("PROJECTION_INVALID",
                        "assertion bound to different action")
    if source_blob["data"] is None:
        raise BondError("HASH_MISMATCH", "evidence source bytes")
    source_bytes = b64u_decode(source_blob["data"])
    src = signed_assertion["body"]["source"]
    if H(source_bytes) != src["hash"] or len(source_bytes) != src["bytes"]:
        raise BondError("HASH_MISMATCH", "evidence source bytes")
    verify_source(signed_assertion["body"]["facts"], signed_assertion["body"],
                  source_bytes, trust, tenant_id, source_verifiers)
    evidence_blob = json_blob(signed_assertion)
    return {
        "assertion": signed_assertion,
        "facts": signed_assertion["body"]["facts"],
        "source_bytes": source_bytes,
        "evidence_ref": ref_of(evidence_blob),
        "evidence_blob": evidence_blob,
    }


def make_fixture_evidence(facts: dict, action_hash: str, observed_at: str,
                          key_id: str, signer: Signer) -> dict:
    """Construct simulation evidence (fixture/1)."""
    source = json_blob(facts)
    assertion = sign_object("assertion", {
        "schema": "bond.assertion/1",
        "action_hash": action_hash,
        "observed_at": observed_at,
        "source_profile": "fixture/1",
        "source": ref_of(source),
        "facts": facts,
    }, key_id, signer)
    return {"assertion": assertion, "source": source}


def parse_assertion_blob(data: bytes) -> dict:
    return v_signed(parse_json_bytes(data, PACKAGE_LIMITS), "assertion", v_assertion)

File: packages/core/src/evidence.ts
/**
 * Evidence verification and construction (§3.3, §3.5). An evidence artifact is
 * a signed Assertion whose `source` ref pins the native source bytes. The
 * supported source profiles in OSS are `fixture/1` (source bytes = J(facts))
 * and `bond.draft-store/1` (source bytes = J(Signed<DraftStoreRecord>) under
 * the draft-store domain). Native upstream profiles require an installed
 * profile verifier; without one the result is UNSUPPORTED_COMPOSITION.
 */

import { BondError } from "./errors.js";
import { H, J, deepEqual, parseJsonBytes, PACKAGE_LIMITS } from "./canon.js";
import { vAssertion, vDraftStoreRecord, vSigned, vBlob } from "./schema.js";
import { verifySigned, Signer, signObject } from "./sign.js";
import * as T from "./types.js";

export interface VerifiedEvidence {
  assertion: T.Signed<T.Assertion>;
  facts: T.Facts;
  sourceBytes: Buffer;
  /** ArtifactRef of the assertion blob (what event data references). */
  evidenceRef: T.ArtifactRef;
  /** Canonical assertion blob stored in objects/packages. */
  evidenceBlob: T.Blob;
}

export type BlobStore = Map<string, Buffer>;

export function blobOf(bytes: Buffer, mediaType: T.ArtifactRef["media_type"]): T.Blob {
  return {
    hash: H(bytes),
    bytes: bytes.length,
    media_type: mediaType,
    data: Buffer.from(bytes).toString("base64url"),
  };
}

export function refOf(b: T.Blob | { hash: string; bytes: number; media_type: string }): T.ArtifactRef {
  return { hash: b.hash, bytes: b.bytes, media_type: b.media_type as T.ArtifactRef["media_type"] };
}

export function jsonBlob(x: unknown): T.Blob {
  return blobOf(Buffer.from(J(x), "utf8"), "application/json");
}

/** Native source-profile verifier installed by an adapter. */
export type SourceVerifier = (
  facts: T.Facts,
  sourceBytes: Buffer,
  assertion: T.Assertion,
  trust: T.TrustFile,
  tenantId: string,
) => void;

export interface VerifyOptions {
  sourceVerifiers?: Partial<Record<T.SourceProfile, SourceVerifier>>;
}

/**
 * Verify the native source bytes against the asserted facts. `fixture/1`
 * requires bytes === J(facts); `bond.draft-store/1` requires a signed
 * DraftStoreRecord whose facts equal the asserted facts.
 */
export function verifySource(
  facts: T.Facts,
  assertion: T.Assertion,
  sourceBytes: Buffer,
  trust: T.TrustFile,
  tenantId: string,
  opts: VerifyOptions = {},
): void {
  switch (assertion.source_profile) {
    case "fixture/1":
      if (J(facts) !== sourceBytes.toString("utf8")) {
        throw new BondError("PROJECTION_INVALID", "fixture source does not match facts");
      }
      return;
    case "bond.draft-store/1": {
      const parsed = vSigned(parseJsonBytes(sourceBytes, PACKAGE_LIMITS), "draft-store", vDraftStoreRecord);
      verifySigned("draft-store", parsed, trust, tenantId);
      if (parsed.body.action_hash !== assertion.action_hash ||
        !deepEqual(parsed.body.facts, facts)) {
        throw new BondError("PROJECTION_INVALID", "draft-store record does not match assertion");
      }
      return;
    }
    default: {
      const v = opts.sourceVerifiers?.[assertion.source_profile];
      if (!v) {
        throw new BondError(
          "UNSUPPORTED_COMPOSITION",
          `no verifier installed for source profile ${assertion.source_profile}`,
        );
      }
      v(facts, sourceBytes, assertion, trust, tenantId);
    }
  }
}

/**
 * Full evidence verification against an action hash and a blob store.
 * Returns the verified evidence, including the canonical assertion blob ref
 * that entries must reference.
 */
export function verifyEvidence(
  signedAssertion: T.Signed<T.Assertion>,
  sourceBlob: T.Blob,
  trust: T.TrustFile,
  tenantId: string,
  expectedActionHash: string,
  opts: VerifyOptions = {},
): VerifiedEvidence {
  verifySigned("assertion", signedAssertion, trust, tenantId);
  if (signedAssertion.body.action_hash !== expectedActionHash) {
    throw new BondError("PROJECTION_INVALID", "assertion bound to different action");
  }
  const sourceBytes = Buffer.from(sourceBlob.data ?? "", "base64url");
  if (sourceBlob.data === null || H(sourceBytes) !== signedAssertion.body.source.hash ||
    sourceBytes.length !== signedAssertion.body.source.bytes) {
    throw new BondError("HASH_MISMATCH", "evidence source bytes");
  }
  verifySource(signedAssertion.body.facts, signedAssertion.body, sourceBytes, trust, tenantId, opts);
  const evidenceBlob = jsonBlob(signedAssertion);
  return {
    assertion: signedAssertion,
    facts: signedAssertion.body.facts,
    sourceBytes,
    evidenceRef: refOf(evidenceBlob),
    evidenceBlob,
  };
}

/**
 * Construct simulation evidence (fixture/1): source bytes are J(facts), the
 * assertion is signed under the "assertion" domain.
 */
export function makeFixtureEvidence(
  facts: T.Facts,
  actionHash: string,
  observedAt: string,
  keyId: string,
  signer: Signer,
): T.Evidence {
  const source = jsonBlob(facts);
  const assertion = signObject("assertion", {
    schema: "bond.assertion/1",
    action_hash: actionHash,
    observed_at: observedAt,
    source_profile: "fixture/1",
    source: refOf(source),
    facts,
  } as T.Assertion, keyId, signer);
  return { assertion, source };
}

/**
 * Construct an installed draft-store observation: the source is a Signed
 * DraftStoreRecord under the "draft-store" domain; the assertion carries
 * source_profile bond.draft-store/1.
 */
export function makeDraftStoreEvidence(
  record: Omit<T.DraftStoreRecord, "schema">,
  storeKeyId: string,
  storeSigner: Signer,
  assertionKeyId: string,
  assertionSigner: Signer,
): T.Evidence {
  const signedRecord = signObject("draft-store", {
    schema: "bond.draft-store/1",
    ...record,
  } as T.DraftStoreRecord, storeKeyId, storeSigner);
  const source = jsonBlob(signedRecord);
  const assertion = signObject("assertion", {
    schema: "bond.assertion/1",
    action_hash: record.action_hash,
    observed_at: record.observed_at,
    source_profile: "bond.draft-store/1",
    source: refOf(source),
    facts: record.facts,
  } as T.Assertion, assertionKeyId, assertionSigner);
  return { assertion, source };
}

/** Parse + validate a signed assertion from stored/package bytes. */
export function parseAssertionBlob(bytes: Buffer): T.Signed<T.Assertion> {
  return vSigned(parseJsonBytes(bytes, PACKAGE_LIMITS), "assertion", vAssertion);
}

export function parseBlob(b: T.Blob): Buffer {
  if (b.data === null) throw new BondError("INCOMPLETE", "blob bytes absent");
  const raw = Buffer.from(b.data, "base64url");
  if (raw.length !== b.bytes || H(raw) !== b.hash) {
    throw new BondError("HASH_MISMATCH", "blob content");
  }
  return raw;
}

export { vBlob };
